### DISCUSSION TITLE
Issue with Handling Image Fields When Updating Records

### DISCUSSION THREAD

**Discussion by FrancineHuang** [aminalaee/sqladmin#774] [Q&A] [answered]
Hi,

I'm back with another question regarding the issue with handling image fields when updating records in sqladmin.

**Issue Description**
When I try to update a record that includes an image field, if no new image is uploaded, the existing path should remain unchanged. (I have already write a logic to upload image on S3 bucket and store the image path on database)

However, I face an AttributeError indicating that a 'str' object has no attribute 'name'. This occurs because the library seems to attempt creating a new UploadFile object from a string path (the existing image URL) when no file is uploaded.

**Code Example**
In my `update_model` method, I check if there's a new file uploaded and process it. If no new file is uploaded, I skip updating the image path. Here’s a simplified version of my implementation:

```
async def update_model(self, request: Request, pk, data: dict):
    form = await request.form()
    image_data = form.get("image")
    if image_data and image_data.filename:
        # Process new image upload
        # (Code to upload new image and update path in the database)
    else:
        # No new image uploaded, retain existing image path
```

**Error Message**

```
 File "{library_path}/site-packages/sqladmin/application.py", line 689, in _handle_form_data
          form_data.append((key, UploadFile(filename=f.name, file=f.open())))
                                                     ^^^^^^
      AttributeError: 'str' object has no attribute 'name'
```
This error occurs during the internal handling of form data when no new image file is uploaded, as the system still tries to process the image field as if a new file should be there.

**Questions**

1. Is there a recommended way to handle such cases where no new file is uploaded but other fields of the record are being updated?
2. Does sqladmin provide configuration options or methods to bypass the automatic file handling for updates where no new file is provided?

Thanks for your assistance🙏

**Comment by aminalaee** [comment]
Hey, what version of sqladmin are you using?

**Reply by FrancineHuang** [reply]
@aminalaee
Hey, sorry for my lately reply. My current version of sqladmin is `sqladmin = {extras = ["full"], version = "^0.16.1"}`

**Reply by aminalaee** [reply]
I think this was fixed in version 0.17.0

**Reply by FrancineHuang** [reply]
Actually...after receiving your suggestion, I updated sqladmin to the latest version (0.17.0).
However, the error still occurs when I try to update a text field without uploading any image file: AttributeError: 'str' object has no attribute 'name'.
Could this issue be related to the fact that I'm using my custom S3 upload logic instead of the fastapi-storages library?

**Additional info**
Here is the definition of the image column in my model:
```
image = Column(String(255), nullable=True)
```

And here is how I store the data into the image column. I need to store a generated path for the S3 filesystem:
```
uploads/images/posts/(image-uuid).webp
```

However, I am encountering issues with the following code from the library:
```
    async def _handle_form_data(self, request: Request, obj: Any = None) -> FormData:
        """
        Handle form data and modify in case of UploadFile.
        This is needed since in edit page
        there's no way to show current file of object.
        """

        form = await request.form()
        form_data: List[Tuple[str, Union[str, UploadFile]]] = []
        for key, value in form.multi_items():
            if not isinstance(value, UploadFile):
                form_data.append((key, value))
                continue

            should_clear = form.get(key + "_checkbox")
            empty_upload = len(await value.read(1)) != 1
            await value.seek(0)
            if should_clear:
                form_data.append((key, UploadFile(io.BytesIO(b""))))
            elif empty_upload and obj and getattr(obj, key):
                f = getattr(obj, key)  # In case of update, imitate UploadFile
                form_data.append((key, UploadFile(filename=f.name, file=f.open())))
            else:
                form_data.append((key, value))
        return FormData(form_data)
```

The error message was still the same. This is all of my evidence.

**Reply by aminalaee** [reply, marked as answer]
Yeah, the logic is specifically done for fastapi-storages. If you feel like it can be generalized, of course that'd be great.

**Reply by FrancineHuang** [reply]
Yep, exactly! I am thinking about using fastapi-storages and give it a try. Thanks for your reply:)
